### PR TITLE
Rendering: Optimization & Fixes

### PR DESCRIPTION
**Pull request by maximumpower55**
~~Sorry for the weird git history~~
This PR fixes 2 major transparency issues with text blocks and address 2 todos with baked block entity rendering a/k/a
- Re-use `BufferBuilder`s
- Move rebuilder creation for invalid regions before `// Iterate over all RegionBuilders, render and upload to RegionBuffers`

I don't have exact statistics but this PR does provide a good performance improvement with 100 text blocks all with the backdrop shadow enabled all spaced out by 3 blocks (3 blocks since the baked block entity rendering's "regions" are 2x2 blocks in size).

I will probably been cleaning up some of this PR later today since most of it was coded at 1 am lol

This description is a bit outdated, too lazy to write out a new one right now.

### CODE REVIEW

File: src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package dev.hephaestus.glowcase.client;
+
+import net.minecraft.client.render.RenderLayer;
+import net.minecraft.client.render.RenderPhase;
+import net.minecraft.client.render.VertexFormat;
+import net.minecraft.client.render.VertexFormats;
+
+public class GlowcaseRenderLayers extends RenderPhase {
+	// Use a custom render layer to render the text plate - mimics DrawableHelper's RenderSystcallsem
```
Comment (exaskye) [inline src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java]: `RenderSystcallsem`?

File: src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java [hunk 2]
```diff
@@ -194,103 +183,134 @@ public VertexConsumer getBuffer(RenderLayer layer) {
...
-					if (currentWorld == null) {
-						break;
-					}
+			Vec3d cam = wrc.camera().getPos();
 
-					for (int chunkX = rrp.x << REGION_FROMCHUNK_SHIFT; chunkX < (rrp.x + 1) << REGION_FROMCHUNK_SHIFT; chunkX++) {
-						for (int chunkZ = rrp.z << REGION_FROMCHUNK_SHIFT; chunkZ < (rrp.z + 1) << REGION_FROMCHUNK_SHIFT; chunkZ++) {
-							blockEntities.addAll(currentWorld.getChunk(chunkX, chunkZ).getBlockEntities().values());
+			if (!needsRebuild.isEmpty()) {
+				wrc.profiler().swap("rebuild");
```
Comment (exaskye) [inline src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: This should probably be a `push`, not `swap`, so the path is `glowcase:based_block_entity_rendering.rebuild`

If changed, a `pop` needs to be added before this if exits

File: src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java [hunk 3]
```diff
@@ -194,103 +183,134 @@ public VertexConsumer getBuffer(RenderLayer layer) {
...
-						BufferBuilder builder = layerBuilder.getValue();
+						RegionBuffer buf = regions.computeIfAbsent(rrp, k -> new RegionBuffer());
+						for (Map.Entry<RenderLayer, BufferBuilder> layerBuilder : entryBuilder.getValue()) {
+							RenderLayer layer = layerBuilder.getKey();
+							BufferBuilder builder = layerBuilder.getValue();
 
-						if (!builder.isBuilding()) builder.begin(layer.getDrawMode(), layer.getVertexFormat());
-						buf.upload(layer, builder);
+							if (!builder.isBuilding()) builder.begin(layer.getDrawMode(), layer.getVertexFormat());
+							buf.upload(layer, builder);
```
Comment (exaskye) [inline src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: Why do you start building only to immediately upload it? The presence of a `isBuilding` guard is a code smell, you should already know the state of your builders...

File: src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java (hunk 1 above)
Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java]: Oops, that supposed to say `RenderSystem call`

File: src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java (hunk 2 above)
Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: 👍

File: src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java (hunk 3 above)
Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: 1: I start building it right before uploading so the VBO is empty and does not contain the data from the last rebuild
2: The `isBuilding` guard is needed so when you make a text block empty it doesn't crash

Please correct me if I am wrong with any of these answers

Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: Removed the code smell, added proper fixes to what it was trying to fix.

File: src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java (hunk 2 above)
Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/render/block/entity/BakedBlockEntityRenderer.java]: Done.

File: src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java (hunk 1 above)
Comment (maximumpower55) [reply src/main/java/dev/hephaestus/glowcase/client/GlowcaseRenderLayers.java]: Fixed.

**Pull request merged**